### PR TITLE
Fix #1024 - fix segfault caused by incorrect hashmap iteration 

### LINKED ISSUE #1024: Properly fix hashmap invalidation issue
This PR implemented a workaround: https://github.com/bpfkorea/agora/pull/1023

However it is inefficient, we should not use an array for all the list of clients since the array may grow too much (although we could put limitations in place, and remove unused clients).

Here is a test-case showing the issue:

```D
import std.stdio;
import agora.utils.Test;
import agora.common.crypto.Key;

void test () @safe
{
    int[PublicKey] map;
    map[WK.Keys[0].address] = 10;
    map[WK.Keys[1].address] = 20;
    map[WK.Keys[2].address] = 30;
    map[WK.Keys[3].address] = 40;
    map[WK.Keys[4].address] = 50;
    size_t count;
    foreach (ref key, ref value; map)
    {
        writeln("In map (before blowItUp) Key: ", key, ", value: ", value);
        ++count;
        if (count == 3)
            blowItUp(map);
        writeln("In map (after blowItUp) Key: ", key, ", value: ", value);
    }
}

void blowItUp (ref int[PublicKey] map) @safe
{
    for (int v = 6; v < 512; ++v)
    {
        map[WK.Keys[v].address] = v * 10;
        writeln("Adding: Key: ", WK.Keys[v].address, ", value: ", v * 10);
    }
}

unittest
{
    test();
}
```

Notice the corruption here in the printouts:

```
n map (after blowItUp) Key: GDC22CFFKB4ZNRZUP6EMRIGVZSQEPSNH2CBMWLU5GLGKE36M3KX5YD36, value: 30
In map (before blowItUp) Key: GAEAAAAAAAAAAAGPHEVAWAIAAAAAAAAAAAAAAAAAIAA5SCQBAAAABOWV, value: 10
In map (after blowItUp) Key: GAEAAAAAAAAAAAGPHEVAWAIAAAAAAAAAAAAAAAAAIAA5SCQBAAAABOWV, value: 10
In map (before blowItUp) Key: GABQAAAAAAAAAAFKHEVAWAIAAAAAAAAAAAAAAAAAUD5NQCQBAAAAB2H7, value: 10
In map (after blowItUp) Key: GABQAAAAAAAAAAFKHEVAWAIAAAAAAAAAAAAAAAAAUD5NQCQBAAAAB2H7, value: 10
In map (before blowItUp) Key: GAFAAAAAAAAAAAGEHEVAWAIAAAAAAAAAAAAAAAAAUD65QCQBAAAAAYG6, value: 8
In map (after blowItUp) Key: GAFAAAAAAAAAAAGEHEVAWAIAAAAAAAAAAAAAAAAAUD65QCQBAAAAAYG6, value: 8
In map (before blowItUp) Key: GAFAAAAAAAAAAAGYHEVAWAIAAAAAAAAAAAAAAAAA6AA5SCQBAAAAAMIZ, value: 1919903585
In map (after blowItUp) Key: GAFAAAAAAAAAAAGYHEVAWAIAAAAAAAAAAAAAAAAA6AA5SCQBAAAAAMIZ, value: 1919903585
```

At first I suspected it was a call to `_aaRehash`, however this is actually the user-facing API for the `.rehash()` call (`__aaRehash` will not be called automatically). There is probably an internal call to one of the functions here: https://github.com/dlang/druntime/blob/master/src/rt/aaA.d#L126-L155

But I haven't debugged it yet.

### PR DESCRIPTION
**Pull request by ferencdg**
issue #1024: fix segfault caused by incorrect hashmap iteration with addition

### CODE REVIEW

File: source/agora/network/NetworkManager.d [hunk 1]
```diff
@@ -792,7 +794,7 @@ public class NetworkManager
 
     public void sendPreimage (PreImageInfo preimage) @safe
     {
-        foreach (ref node; this.peers)
+        foreach (ref node; this.peers[])
```
Comment (Geod24) [inline source/agora/network/NetworkManager.d:797]: `DList` does not define `opApply` ?

Comment (Geod24) [reply source/agora/network/NetworkManager.d:797]: Ah, `opSlice` returns a `Range`, how weird.

**Review by Geod24 — APPROVED** [review]
LGTM

**Review by linked0 — APPROVED** [review]
LGTM

**Pull request merged**